File: core/src/main/java/com/lomekwi/cave/pipeline/image/TransNode.java
package com.lomekwi.cave.pipeline.image;

import com.lomekwi.cave.pipeline.Filter;
import com.lomekwi.cave.pipeline.num.NumFrame;

import static com.lomekwi.cave.pipeline.Node.*;

public class TransNode extends Filter<Transformable> {

    private final InPort<NumFrame> dx = addInPort(
        new InPort<>("位移 X", new NumFrame(null), NumFrame.class));
    private final InPort<NumFrame> dy = addInPort(
        new InPort<>("位移 Y", new NumFrame(null), NumFrame.class));
    private final InPort<NumFrame> scaleX = addInPort(
        new InPort<>("缩放 X", frame(1), NumFrame.class));
    private final InPort<NumFrame> scaleY = addInPort(
        new InPort<>("缩放 Y", frame(1), NumFrame.class));
    private final InPort<NumFrame> dRotation = addInPort(
        new InPort<>("旋转", new NumFrame(null), NumFrame.class));

    private boolean flipX, flipY;

    private final FilterIn in = addInPort(new FilterIn("输入") {
    });

    private final FilterOut out = addOutPort(new FilterOut("输出") {
        @Override
        public Transformable getData() {
            Transformable frame = getFilterIn().getData();
            if (frame != null) {
                Transform t = frame.getTransform();
                if (t == null) {
                    t = new Transform();
                    frame.setTransform(t);
                }
                t.applyLocal((float) val(dx), (float) val(dy), (float) val(scaleX), (float) val(scaleY),
                        (float) val(dRotation), flipX, flipY);
            }
            return frame;
        }
    });

    public TransNode() {
    }

    public TransNode(double dx, double dy, double scaleX, double scaleY, double dRotation) {
        setDx(dx);
        setDy(dy);
        setScaleX(scaleX);
        setScaleY(scaleY);
        setDRotation(dRotation);
    }

    private static NumFrame frame(double v) {
        NumFrame f = new NumFrame(null);
        f.setVal(v);
        return f;
    }

    private static double val(InPort<NumFrame> p) {
        return p.getData().getVal();
    }

    public double getDx() {
        return val(dx);
    }

    public void setDx(double v) {
        dx.getDefaultData().setVal(v);
    }

    public double getDy() {
        return val(dy);
    }

    public void setDy(double v) {
        dy.getDefaultData().setVal(v);
    }

    public double getScaleX() {
        return val(scaleX);
    }

    public void setScaleX(double v) {
        scaleX.getDefaultData().setVal(v);
    }

    public double getScaleY() {
        return val(scaleY);
    }

    public void setScaleY(double v) {
        scaleY.getDefaultData().setVal(v);
    }

    public double getDRotation() {
        return val(dRotation);
    }

    public void setDRotation(double v) {
        dRotation.getDefaultData().setVal(v);
    }

    public boolean flipX() {
        return flipX;
    }

    public void flipX(boolean v) {
        flipX = v;
    }

    public boolean flipY() {
        return flipY;
    }

    public void flipY(boolean v) {
        flipY = v;
    }

    @Override
    public Class<Transformable> getType() {
        return Transformable.class;
    }

    @Override
    public String getName() {
        return "变换";
    }
}

File: core/src/main/scala/com/lomekwi/cave/pipeline/image/TransNode.scala
package com.lomekwi.cave.pipeline.image

import com.lomekwi.cave.pipeline.Filter
import com.lomekwi.cave.pipeline.Node

class TransNode extends Filter[Transformable] {

  private final val dx: Node.InPort[Double] = addInPort(
    new Node.InPort[Double]("位移 X", 0.0, classOf[Double]))
  private final val dy: Node.InPort[Double] = addInPort(
    new Node.InPort[Double]("位移 Y", 0.0, classOf[Double]))
  private final val scaleX: Node.InPort[Double] = addInPort(
    new Node.InPort[Double]("缩放 X", 1.0, classOf[Double]))
  private final val scaleY: Node.InPort[Double] = addInPort(
    new Node.InPort[Double]("缩放 Y", 1.0, classOf[Double]))
  private final val dRotation: Node.InPort[Double] = addInPort(
    new Node.InPort[Double]("旋转", 0.0, classOf[Double]))

  private var flipXState: Boolean = false
  private var flipYState: Boolean = false

  private final val in: FilterIn = addInPort(new FilterIn("输入") {
  })

  private final val out: FilterOut = addOutPort(new FilterOut("输出") {
    override def getData: Transformable = {
      val frame: Transformable = getFilterIn.getData
      if (frame != null) {
        var t: Transform = frame.getTransform
        if (t == null) {
          t = new Transform()
          frame.setTransform(t)
        }
        t.applyLocal(TransNode.value(dx).toFloat, TransNode.value(dy).toFloat, TransNode.value(scaleX).toFloat, TransNode.value(scaleY).toFloat,
                TransNode.value(dRotation).toFloat, flipXState, flipYState)
      }
      frame
    }
  })

  def this(dx: Double, dy: Double, scaleX: Double, scaleY: Double, dRotation: Double) = {
    this()
    setDx(dx)
    setDy(dy)
    setScaleX(scaleX)
    setScaleY(scaleY)
    setDRotation(dRotation)
  }

  def getDx: Double = {
    TransNode.value(dx)
  }

  def setDx(v: Double): Unit = {
    dx.setDefaultData(v)
  }

  def getDy: Double = {
    TransNode.value(dy)
  }

  def setDy(v: Double): Unit = {
    dy.setDefaultData(v)
  }

  def getScaleX: Double = {
    TransNode.value(scaleX)
  }

  def setScaleX(v: Double): Unit = {
    scaleX.setDefaultData(v)
  }

  def getScaleY: Double = {
    TransNode.value(scaleY)
  }

  def setScaleY(v: Double): Unit = {
    scaleY.setDefaultData(v)
  }

  def getDRotation: Double = {
    TransNode.value(dRotation)
  }

  def setDRotation(v: Double): Unit = {
    dRotation.setDefaultData(v)
  }

  def flipX(): Boolean = {
    flipXState
  }

  def flipX(v: Boolean): Unit = {
    flipXState = v
  }

  def flipY(): Boolean = {
    flipYState
  }

  def flipY(v: Boolean): Unit = {
    flipYState = v
  }

  override def getName: String = {
    "变换"
  }
}

object TransNode {
  private def value(p: Node.InPort[Double]): Double = {
    p.getData
  }
}
